### PR TITLE
[FLINK-6250] Distinct procTime with Rows boundaries

### PR DESCRIPTION
**Pull request by stefanobortoli**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [X ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ X] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ X] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by stefanobortoli** [comment]
@fhueske @rtudoran @shijinkui @sunjincheng121 I have create a new PR for distinct in the code generator. Please have a look and let me know. I have implemented and tested only for OverProcTimeRowBounded window, but if you like it I can quickly implement and test also the others.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 1]
```diff
@@ -290,6 +295,10 @@ object AggregateUtil {
     val aggMapping = aggregates.indices.toArray.map(_ + groupings.length)
     val outputArity = aggregates.length + groupings.length + 1
 
+    // remove when distinct is supported
+    val distinctAggregatesFlags = new Array[Boolean](aggregates.size)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: `DataSet` has it's own way to deal with distinct aggregates. They could also not initialize the state. So it is save to call `generateAggregations()` with an empty array: `Array[Boolean]()` or we make the parameter an `Option[Array[Boolean]]` and pass `None`

Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: The same applies for all DataSet aggregations.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala [hunk 2]
```diff
@@ -20,16 +20,22 @@ package org.apache.flink.table.runtime.aggregate
...
+  
+  /**
+    * Initialize the state for the distinct aggregation check
+    *
+    * @param ctx the runtime context to retrieve and initialize the distinct states
+    */
+  def initialize(ctx: RuntimeContext)
 
   /**
-    * Sets the results of the aggregations (partial or final) to the output row.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala]: Revert this change?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala [hunk 3]
```diff
@@ -72,7 +73,15 @@ class ProcTimeBoundedRowsOver(
       genAggregations.code)
     LOG.debug("Instantiating AggregateHelper.")
     function = clazz.newInstance()
-
+    
+    var initialized = false
+    for(i <- distinctAggFlags.indices){
+      if(distinctAggFlags(i) && !initialized){
+        function.initialize(getRuntimeContext())
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala]: I would always call `function.initialize()`. If there are no distinct aggregates, nothing will happen

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala [hunk 4]
```diff
@@ -46,6 +46,7 @@ import org.slf4j.LoggerFactory
   */
 class ProcTimeBoundedRowsOver(
     genAggregations: GeneratedAggregationsFunction,
+    distinctAggFlags: Array[Boolean],
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala]: remove this parameter.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRangeProcessFunctionTest.scala [hunk 5]
```diff
@@ -88,6 +88,12 @@ class BoundedProcessingOverRangeProcessFunctionTest {
         |    "mluZyRMb25nJOda0iCPo2ukAgAAeHA");
         |  }
         |
+        |  public void initialize(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRangeProcessFunctionTest.scala]: can you make this a one-line change?

**Comment by huawei-flink** [comment]
@fhueske @haohui I have no problem removing the DIST() part, it is just not possible to test it without. Shall I push just the code generation and aggregates util changes?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 6]
```diff
@@ -296,6 +297,41 @@ class CodeGenerator(
...
+      
+      val sig: String = 
+        j"""
+           | org.apache.flink.api.common.state.MapState[] distStateList =
+           |   new org.apache.flink.api.common.state.MapState[ ${distinctAggsFlags.size} ];
+           | 
+           |  public void initialize(
+           |    org.apache.flink.api.common.functions.RuntimeContext ctx
+           |  )""".stripMargin
+      if(existDistinct){
```
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: you are right

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 7]
```diff
@@ -335,14 +371,28 @@ class CodeGenerator(
...
-            |    org.apache.flink.types.Row input)""".stripMargin
+            |    org.apache.flink.types.Row input) throws Exception""".stripMargin
 
       val accumulate: String = {
         for (i <- aggs.indices) yield
-          j"""
+         if(distinctAggsFlags(i)){
+           j"""
+              |  Long distValCount$i = (Long) distStateList[$i].get(${parameters(i)});
+              |  if( distValCount$i == null){
```
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: In scala it is 0L, is Java it is null. :-/

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala (hunk 2 above)
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala]: probably an error in the merging. sorry about that

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala (hunk 3 above)
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala]: right!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala (hunk 7 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: Ah, that makes sense. Scala `Long` is backed by Java `long`, i.e., it does not support `null`.

**Comment by fhueske** [comment]
If we merge this change together with #3764, we would check it with tests that use distinct grouped window aggregates. 
But you are right, it might in fact make sense to test the `GeneratedAggregations` class generated by the `CodeGenerator` individually in a unit test. I wouldn't make this as part of this PR though.

**Comment by huawei-flink** [comment]
So, what do you want me to keep for this PR? just the code generation and its test?

**Comment by huawei-flink** [comment]
@fhueske I have created #3783 with just the code generation part. At least the GROUP BY distinct can move ahead. I will close this PR and wait for the merging of the Calcite fix.
